hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/codec/NSSummaryCodec.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.recon.codec;
+
+import org.apache.hadoop.hdds.utils.db.IntegerCodec;
+import org.apache.hadoop.ozone.recon.ReconConstants;
+import org.apache.hadoop.ozone.recon.api.types.NSSummary;
+import org.apache.hadoop.hdds.utils.db.Codec;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+
+/**
+ * Codec for Namespace Summary.
+ */
+public class NSSummaryCodec implements Codec<NSSummary>{
+
+  private final Codec<Integer> integerCodec = new IntegerCodec();
+  // 2 int fields + 40-length int array
+  private static final int NUM_OF_INTS = 2 + ReconConstants.NUM_OF_BINS;
+
+  @Override
+  public byte[] toPersistedFormat(NSSummary object) throws IOException {
+    final int sizeOfRes = NUM_OF_INTS * Integer.BYTES;
+    ByteArrayOutputStream out = new ByteArrayOutputStream(sizeOfRes);
+    out.write(integerCodec.toPersistedFormat(object.getNumOfFiles()));
+    out.write(integerCodec.toPersistedFormat(object.getSizeOfFiles()));
+    int[] fileSizeBucket = object.getFileSizeBucket();
+    for (int i = 0; i < ReconConstants.NUM_OF_BINS; ++i) {
+      out.write(integerCodec.toPersistedFormat(fileSizeBucket[i]));
+    }
```

Comment:
One suggestion here: we should probably write NUM_OF_BINS before writing the array. Just in case we expand or shrink the array in the future. Accordingly we need to handle this in `fromPersistedFormat`.

An even better approach would be to have an `IntegerArrayCodec` that handles the dynamic length of int array?

Reply:
Even we expand or shrink the num_of_bins in the future, the current approach should have no problem with it. In `NSSummaryCodec`, I calculated the number of ints stored in NSSummary:
`private static final int NUM_OF_INTS = 2 + ReconConstants.NUM_OF_BINS;`
then, I calculated the number of bytes needed for (de)serialization:
`final int sizeOfRes = NUM_OF_INTS * Integer.BYTES;`
and
`assert(rawData.length == NUM_OF_INTS * Integer.BYTES);`,
so even if in the future we change the num of bins, we don't need any modification in `NSSummaryCodec`

Reply:
hmm that wasn't my point. I'm point is the future upgradability. Consider this: if we expanded the bins by 1 slot in the future. NUM_OF_INTS in the executed binary will be 41+1=42. If we are still reading from the **old** RDB, there won't be a 42-th entry in the byte array. So it'll likely throw an exception at this point.

The worse case is, if we also added a new int field to the NSSummary class at the same time indicating a different metric. Then the new code won't know if the 42-th entry is the new bin or new field. Because the new code in `fromPersistedFormat` is using the new `NUM_OF_INTS=42`, not the old `NUM_OF_INTS=41`.

The solution is very simple. Just writing an extra int indicating _the length of array_ right before the content (list of ints) of the array. Even if we don't do much in `fromPersistedFormat` right now, it will help later on that we add logic to get the correct array length. For example, just read the length-indicating int and throw it away. It will help when expand the DB.

That's also part of the reason why we use protobuf to deal with all those compatibility details.

Reply:
We could also use built-in ArrayList here to not deal with the length of array ourselves. https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/ArrayList.java#L862